### PR TITLE
feat-utopia-db-integration

### PR DESCRIPTION
**Pull request by lohanidamodar**

### CODE REVIEW

**Review by eldadfux — CHANGES_REQUESTED** [review]

File: src/Abuse/Adapters/TimeLimit.php [hunk 1]
```diff
@@ -149,55 +127,72 @@ protected function parseKey(): string
...
+            new Query('_key', Query::TYPE_EQUAL, [$key]),
+            new Query('_time', Query::TYPE_EQUAL, [$time]),
+        ], 1);
+        Authorization::reset();
+
+        if (count($result) == 1) {
+            $result = $result[0]->getAttribute('_count',0);
+        } else {
+            $result = 0;
+        }
```
Comment (kodumbeats) [inline src/Abuse/Adapters/TimeLimit.php:149]: Can this be a ternary operator? We can also typecast `(int) $result` here too.

```php
$this->count = (\count($result) === 1) ? (int) $result[0]->getAttribute('_count', 0) : 0;
```

File: src/Abuse/Adapters/TimeLimit.php [hunk 2]
```diff
@@ -149,55 +127,72 @@ protected function parseKey(): string
...
-    	$result = $st->fetch();
-    	$result = (\is_array($result) && isset($result['_count'])) ? $result['_count'] : 0;
+        Authorization::disable();
+        $result = $this->db->find(TimeLimit::COLLECTION, [
+            new Query('_key', Query::TYPE_EQUAL, [$key]),
+            new Query('_time', Query::TYPE_EQUAL, [$time]),
+        ], 1);
+        Authorization::reset();
+
+        if (count($result) == 1) {
```
Comment (kodumbeats) [inline src/Abuse/Adapters/TimeLimit.php:145]: Can we clearly distinguish between `Timelimit->count()` and PHP's standard `count()` and check strict equality:
```suggestion
        if (\count($result) === 1) {
```

File: src/Abuse/Adapters/TimeLimit.php [hunk 3]
```diff
@@ -149,55 +127,72 @@ protected function parseKey(): string
...
+        $data = [
+            '$read' => [],
+            '$write' => [],
+            '_key' => $key,
+            '_time' => $time,
+            '_count' => 1,
+            '$collection' => TimeLimit::COLLECTION,
+        ];
+
+        if (count($existing) == 1) {
```
Comment (kodumbeats) [inline src/Abuse/Adapters/TimeLimit.php:185]:
```suggestion
        if (\count($existing) === 1) {
```

File: src/Abuse/Adapters/TimeLimit.php [hunk 4]
```diff
@@ -207,45 +202,41 @@ protected function hit(string $key, int $time): void
...
+     * @return bool
      */
-    public function cleanup(int $timestamp):bool
+    public function cleanup(int $timestamp): bool
     {
-        $st = $this->getPDO()->prepare('DELETE 
-        FROM `'.$this->getNamespace().'.abuse.abuse`
-            WHERE `_time` < :timestamp');
+        Authorization::disable();
+        $documents = $this->db->find(TimeLimit::COLLECTION, [
```
Comment (kodumbeats) [inline src/Abuse/Adapters/TimeLimit.php:228]: Will this actually clean up all necessary documents? `\Utopia\Database\Database->find()` has a default limit of 25 records:
https://github.com/utopia-php/database/blob/main/src/Database/Database.php#L813

**Review by kodumbeats — CHANGES_REQUESTED** [review]
:)

File: src/Abuse/Adapters/TimeLimit.php (hunk 4 above)
Comment (eldadfux) [reply src/Abuse/Adapters/TimeLimit.php:228]: Good catch, we should iterate the results until no results come up - similar to what we do on the deletes worker.

File: src/Abuse/Adapters/TimeLimit.php [hunk 5]
```diff
@@ -207,45 +202,43 @@ protected function hit(string $key, int $time): void
...
+        Authorization::disable();
+        do {
+            $documents = $this->db->find(TimeLimit::COLLECTION, [
+                new Query('time', Query::TYPE_LESSER, [$timestamp]),
+            ]);
+    
+            foreach ($documents as $document) {
+                $this->db->deleteDocument(TimeLimit::COLLECTION, $document['$id']);
+            }
+        } while(\count($documents) > 0);
```
Comment (eldadfux) [inline src/Abuse/Adapters/TimeLimit.php:236]: Don't use `count`, `!empty` should be more efficient

**Review by eldadfux — APPROVED** [review]

**Review by kodumbeats — APPROVED** [review]

**Pull request merged**